rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,476 @@
+other components.
+
+Example: Change the border radius and border width of all Button components without changing those styles for
+Input components.
+
+## Option A: React.Context useCustomStyles hooks
+
+Each component could check context for a useCustomStyle hook within the useComponentStyle hook.
+If this hook was defined, it is called at the end of the method before returning state.
+
+> These example use Component as a substitute for a component like Button or Avatar.
+
+```tsx
+const customStyles = useContextSelector(ComponentCustomStylesContext, context => context?.useCustomStyles);
+
+//...component styling here
+
+return customStyles ? customStyles(state) : state;
+```
+
+Each component defines own types for props and state, so would need to define its own context value types,
+context, and context provider.
+
+```tsx
+export type ComponentCustomStylesContextValue = {
+  useCustomStyles?: (state: ComponentState) => ComponentState;
+};
+
+export const ComponentCustomStylesContext: Context<ComponentCustomStylesContextValue> = createContext<ComponentCustomStylesContextValue>(
+  {},
+);
+
+export const ComponentCustomStylesContextProvider = ComponentCustomStylesContext.Provider;
+```
+
+When a caller wanted to customize the component, they use the provider. Callers would be able to control custom
+styling scope just like Fluent Provider. While they would need to use a provider per custom styled component,
+these could be aggregated into a component that transcludes children.
+
+```tsx
+<App>
+  <FluentProvider theme={webLightTheme}>
+    <ComponentCustomStylesContextProvider value={{ useCustomStyles }}>
+      {/* other component hierarchy */}
+      <Component>Hello custom styles!</Component>
+    </ComponentCustomStylesContextProvider>
+  </FluentProvider>
+</App>
+```
+
+### 👍
+
+- Custom styling gets the same power as the default component useStyles hook.
+- This builds on using slots as the addressable items for applying styles.
+- The custom styles can be built using Griffel with build-time optimization.
+- Can be applied at different scope like FluentProvider.
+- Does not require composing or introducing a new component type that must be adopted in lieu of the default component.
+- Fewer CSS specificity problems as styles can get overwritten.
+- Provides the type-safety expected from Fluent UI React.
+- Minimal code to add to each component. Could be put into the code generation template.
+- Does not significantly affect rendering performance.
+- Partners building components can follow the same practice to support custom styling in their libraries.
+
+### 👎
+
+- Custom styling all components requires adding and configuring lots of context providers.
+- Requires writing new code for every component, adding a story, and documenting the feature.
+- Context is harder to discover as it is outside the props contract.
+- Callers doing custom styling may still take a dependency on the CSS internals of the component if they don't replace a
```

Comment:
They can also break existing style overrides that components in a tree have. It could be a significant problem for HVCs that are embedded into apps: styles work in ProductA, but don't work in the same way in ProductB.

Reply:
Yeah - we will need to separately determine what our policy is on the CSS being an API that we consider to take breaking changes.